### PR TITLE
Implement Program Hash

### PR DESCRIPTION
**Pull request by jelilat**
## Description

This PR adds a function to compute the program hash computation

## Changes:

- Adds the program hash computation

## Types of changes

- New feature (non-breaking change which adds functionality)

## Testing

**Requires testing**: Yes

**Did you write tests??**: Yes

### CODE REVIEW

File: clients/programhash_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+var (
+	//go:embed testdata/class_genesis.json
+	classGenesisBytes []byte
+	//go:embed testdata/class_0_8.json
+	classCairo08Bytes []byte
+	//go:embed testdata/class_0_10.json
+	classCairo10Bytes []byte
```
Comment (IronGauntlets) [inline clients/programhash_test.go:18]: Please add links to where this data was retrieved from

**Comment by omerfirmak** [comment]
I think the way we generate the JSON is so brittle that it would be so cumbersome to maintain. I gave a shot at this one as well and made some progress but it is not complete.

check out `MarshalJSONWithSeparators` at https://github.com/NethermindEth/juno/commit/385474075d64c29d871418e7c438b3528e406645

**Comment by jelilat** [comment]
> I think the way we generate the JSON is so brittle that it would be really cumbersome to maintain. I gave a shot at this one as well and made some progress but it is not complete.
> 
> check out `MarshalJSONWithSeparators` at [3854740](https://github.com/NethermindEth/juno/commit/385474075d64c29d871418e7c438b3528e406645)

I absolutely agree and have been experimenting with different things. Thanks for your update. Trying something new that could further simplify the implementation as well. Hopefully, it works.

**Comment by jelilat** [comment]
@omerfirmak I just made some improvements to `MarshalsJSON`. Please take a look and check for further improvements.

File: clients/programhash.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	contractCodeString := string(contractCodeBytes)
+	hintsIndex := strings.Index(contractCodeString, `"hints":`)            // find location of "hints" in contractCodeString
+	identifierIndex := strings.Index(contractCodeString, `"identifiers":`) // find location of "identifiers" in contractCodeString
+	// replace json "hints" with reformatted "hints"
+	contractCodeString = contractCodeString[:(hintsIndex+8)] + newHint + contractCodeString[identifierIndex-1:]
+	// reformat contractCodeString with
+	contractCodeString = strings.ReplaceAll(contractCodeString, `,`, `, `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `,  `, `, `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `:`, `: `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `:  `, `: `)
```
Comment (omerfirmak) [inline clients/programhash.go:89]: These could easily cause problems as those characters could be used anywhere, not just as delimiters. Lets merge and see if this ends up being a problem.

also we should use escaped whitespace characters (`\r`, `\n`, `\t` etc), it is not easily recognizable what the whitespaces are in string constants.

**Review by omerfirmak — APPROVED** [review]
We should push for standardization of the program JSON. This is just unnecessary amount of work for something so simple.

**Pull request merged**